File: frontend/lib/models/pronostico.dart
/// Pronóstico 1X2 para un partido normal.
enum Signo {
  uno('1'),
  x('X'),
  dos('2');

  final String valorApi;
  const Signo(this.valorApi);

  static Signo? fromApi(String? valor) {
    if (valor == null) return null;
    return Signo.values.firstWhere((s) => s.valorApi == valor);
  }
}

/// Pronóstico de goles del Pleno al 15 (0, 1, 2 o M = 3 o más).
enum Goles {
  cero('0'),
  uno('1'),
  dos('2'),
  m('M');

  final String valorApi;
  const Goles(this.valorApi);

  static Goles? fromApi(String? valor) {
    if (valor == null) return null;
    return Goles.values.firstWhere((g) => g.valorApi == valor);
  }
}

class Pronostico {
  final int? id;
  final int partidoId;
  final Signo? signo;
  final Goles? pleno15Local;
  final Goles? pleno15Visitante;
  final bool? acertado;

  Pronostico({
    this.id,
    required this.partidoId,
    this.signo,
    this.pleno15Local,
    this.pleno15Visitante,
    this.acertado,
  });

  Pronostico copyWith({Signo? signo, Goles? pleno15Local, Goles? pleno15Visitante}) {
    return Pronostico(
      id: id,
      partidoId: partidoId,
      signo: signo ?? this.signo,
      pleno15Local: pleno15Local ?? this.pleno15Local,
      pleno15Visitante: pleno15Visitante ?? this.pleno15Visitante,
      acertado: acertado,
    );
  }

  factory Pronostico.fromJson(Map<String, dynamic> json) {
    return Pronostico(
      id: json['id'] as int?,
      partidoId: json['partido_id'] as int,
      signo: Signo.fromApi(json['signo'] as String?),
      pleno15Local: Goles.fromApi(json['pleno15_local'] as String?),
      pleno15Visitante: Goles.fromApi(json['pleno15_visitante'] as String?),
      acertado: json['acertado'] as bool?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'partido_id': partidoId,
      'signo': signo?.valorApi,
      'pleno15_local': pleno15Local?.valorApi,
      'pleno15_visitante': pleno15Visitante?.valorApi,
    };
  }
}

File: backend/app/models/pronostico.py
from __future__ import annotations

from typing import TYPE_CHECKING

from sqlalchemy import Enum as SAEnum
from sqlalchemy import ForeignKey, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base_class import Base, TimestampMixin
from app.models.enums import GolesEnum, SignoEnum

if TYPE_CHECKING:
    from app.models.columna import Columna
    from app.models.partido import Partido


class Pronostico(Base, TimestampMixin):
    """Pronóstico de un usuario (columna) para un partido concreto de una jornada."""

    __tablename__ = "pronosticos"
    __table_args__ = (UniqueConstraint("columna_id", "partido_id", name="uq_pronostico_columna_partido"),)

    id: Mapped[int] = mapped_column(primary_key=True)
    columna_id: Mapped[int] = mapped_column(ForeignKey("columnas.id", ondelete="CASCADE"), nullable=False)
    partido_id: Mapped[int] = mapped_column(ForeignKey("partidos.id", ondelete="CASCADE"), nullable=False)

    # Pronóstico 1X2, usado en los 14 partidos normales (y opcionalmente en el 15 para columnas no-elige8).
    signo: Mapped[SignoEnum | None] = mapped_column(SAEnum(SignoEnum, name="signo_enum"), nullable=True)

    # Pronóstico de goles del Pleno al 15, exclusivo de la columna Elige 8.
    pleno15_local: Mapped[GolesEnum | None] = mapped_column(SAEnum(GolesEnum, name="goles_enum"), nullable=True)
    pleno15_visitante: Mapped[GolesEnum | None] = mapped_column(SAEnum(GolesEnum, name="goles_enum"), nullable=True)

    columna: Mapped["Columna"] = relationship(back_populates="pronosticos")
    partido: Mapped["Partido"] = relationship(back_populates="pronosticos")

    def __repr__(self) -> str:
        return f"<Pronostico columna_id={self.columna_id} partido_id={self.partido_id} signo={self.signo}>"
